senders/test_server_test.go

```diff
@@ -0,0 +1,73 @@
+package senders
+
+import (
+	"bufio"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"net/http"
+)
+
+func newTestServer(port int) *testServer {
+	metricLines := make([]string, 0)
+	return &testServer{
+		port:        port,
+		metricLines: metricLines,
+	}
+}
+
+type testServer struct {
+	port        int
+	metricLines []string
+	httpServer  *http.Server
+}
+
+func (server *testServer) start() {
+	addr := fmt.Sprintf(":%v", server.port)
+	handler := http.NewServeMux()
+	handler.HandleFunc("/", server.handler())
+	httpServer := &http.Server{Addr: addr, Handler: handler}
+
+	server.httpServer = httpServer
+	go func() {
+		if err := httpServer.ListenAndServe(); err != nil {
+			// handle err
+		}
+	}()
+}
+
+func (server *testServer) handler() func(writer http.ResponseWriter, request *http.Request) {
+	return func(writer http.ResponseWriter, request *http.Request) {
+		newLines, err := decodeMetricLines(request)
+		if err != nil {
+			writer.WriteHeader(500)
+		}
+		for _, metricLine := range newLines {
+			server.metricLines = append(server.metricLines, metricLine)
+		}
+		writer.WriteHeader(200)
+	}
+}
+
+func decodeMetricLines(request *http.Request) ([]string, error) {
+	metricLines := make([]string, 0)
+	reader, err := gzip.NewReader(request.Body)
+	if err != nil {
+		reader.Close()
```

Comment:
Yeah, certainly looking at the way `gzip.NewReader` is implemented now, this would be nil. I also realized that this should return early with the error, so I'll fix that.